---
title: "Comparison of STAU1 hiCLIP hybrids with PARIS hybrids"
author: "Ira Iosub"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  html_notebook:
    toc: yes
    toc_depth: 3
    toc_float: yes
    theme: united
    highlight: tango
    df_print: paged
    code_folding: show
---

#### Libraries
```{r include=FALSE}

library(data.table)
library(rtracklayer)
library(primavera)
library(tidyverse)
library(stringr)
library(paletteer)
library(ggpubr)
library(Biostrings)
theme_set(theme_bw() +
            theme(legend.position = "top"))
```

#### Functions
```{r}
plot_stacked_barchart <- function(data.df, column, group) {
  
  data_counts.df <- data.df %>%
    select(name, {{group}}, {{column}}) %>%
    group_by({{group}}, {{column}}) %>%
    summarise(counts = n()) %>%
    arrange(desc({{column}})) %>%
    mutate(percentage = counts*100 / sum(counts)) %>%
    mutate(pos = cumsum(percentage) - percentage/2)
  
  bar = ggplot() + geom_bar(aes(y = percentage, x = "", fill = {{column}}), alpha = 0.8, data = data_counts.df,
                            stat="identity", width = 0.5) +
    geom_text(data = data_counts.df, aes(x = "", y = pos, label = paste0(round(percentage,1),"%")), size=3) +
    facet_grid(cols = vars({{group}})) +
    #facet_wrap(. ~ Experiment, scales = "free") +
    scale_fill_paletteer_d("rcartocolor::Earth", direction = -1) +
    theme(legend.position="right", legend.direction="vertical",
          legend.title = element_blank()) +
    theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
    ylab("Percentage") +
    xlab("Experiment")
  return(bar)
  
}
```


```{r}
get_nucleotide_frequency <- function(duplexes.df) {
  
  # L_seq <- DNAStringSet(duplexes.df$L_sequence)
  # R_seq <- DNAStringSet(duplexes.df$R_sequence)
  L_seq <- DNAStringSet(duplexes.df$l_paired_residues)
  R_seq <- DNAStringSet(duplexes.df$r_paired_residues)
  
  L_freq <- as.data.frame(oligonucleotideFrequency(L_seq, width = 1, as.prob=TRUE))
  R_freq <- as.data.frame(oligonucleotideFrequency(R_seq, width = 1, as.prob=TRUE))
  duplexes.df <- dplyr::bind_cols(duplexes.df, L_freq, R_freq)
  colnames(duplexes.df) = gsub("...43|...44|...45|...46", "_L", colnames(duplexes.df)) # rename columns
  colnames(duplexes.df) = gsub("...47|...48|...49|...50", "_R", colnames(duplexes.df))
  
  # Pur content - calculate max pur L vs. R arm 
  duplexes.df <- duplexes.df %>%
    rowwise() %>% 
    mutate(L_pur = sum(c(A_L, G_L)), # purine total per hybrid
           R_pur = sum(c(A_R, G_R))) %>%
    mutate(max_pur = max(L_pur, R_pur)) %>%
    mutate(arm = case_when((max_pur == L_pur) ~ "L", # assign maximal purine arm
                           TRUE ~ "R")) %>%
    ungroup() 
  
  # reorient duplex frequencies based on maximal purine content (column "arm")
  nuc_freq_max_pur_arm.df <- duplexes.df %>%
    dplyr::filter(arm == "L") %>%
    mutate(A_freq_max_pur_arm = A_L, A_freq_min_pur_arm = A_R,
           G_freq_max_pur_arm = G_L, G_freq_min_pur_arm = G_R,
           C_freq_max_pur_arm = C_L, C_freq_min_pur_arm = C_R,
           U_freq_max_pur_arm = T_L, U_freq_min_pur_arm = T_R)

  nuc_freq_min_pur_arm.df <- duplexes.df %>%
    dplyr::filter(arm == "R") %>%
    mutate(A_freq_max_pur_arm = A_R, A_freq_min_pur_arm = A_L,
           G_freq_max_pur_arm = G_R, G_freq_min_pur_arm = G_L,
           C_freq_max_pur_arm = C_R, C_freq_min_pur_arm = C_L,
           U_freq_max_pur_arm = T_R, U_freq_min_pur_arm = T_L)

  nuc_freq_reordered.df <- rbind(nuc_freq_max_pur_arm.df, nuc_freq_min_pur_arm.df)
  nuc_freq_reordered.df <- nuc_freq_reordered.df %>% arrange(desc(max_pur))
  
  return(nuc_freq_reordered.df)
  
}

```


```{r}
plot_nucleotide_frequency <- function(data) {
  
  # Prepare for plotting
  high_pur.df <- data %>% 
    select(id, A_freq_max_pur_arm, G_freq_max_pur_arm, C_freq_max_pur_arm, U_freq_max_pur_arm) %>%
    dplyr::arrange(G_freq_max_pur_arm)
  
  high_pur.df <- rowid_to_column(high_pur.df)
  long_high_pur.df <- high_pur.df %>% 
    gather(residue, frequency, A_freq_max_pur_arm:U_freq_max_pur_arm)
  long_high_pur.df$arm = "High purine arm"
  
  low_pur.df <- data %>% 
    select(id, A_freq_min_pur_arm, G_freq_min_pur_arm, C_freq_min_pur_arm, U_freq_min_pur_arm) %>%
    dplyr::arrange(desc(U_freq_min_pur_arm))
  low_pur.df <- rowid_to_column(low_pur.df)
  long_low_pur.df <- low_pur.df %>% 
    gather(residue, frequency, A_freq_min_pur_arm:U_freq_min_pur_arm)
  long_low_pur.df$arm = "Low purine arm"
  
  nuc_freq.df <- rbind(long_high_pur.df, long_low_pur.df)
  nuc_freq.df$residue <- str_sub(nuc_freq.df$residue, 1,1) #trim string to get nucleotide (first) letter
  
  # Plot individual nucleotide frequencies
  nuc_freq.gg <- ggplot(nuc_freq.df, aes(x=rowid, y=frequency, color = residue)) +
    geom_smooth(se=F) + coord_flip() +
    facet_grid(cols = vars(arm)) +
    ylab("Frequency") +
    xlab("ID") +
    scale_color_manual(values = c("darkgreen","dodgerblue4","goldenrod3","firebrick"))
  
  return(nuc_freq.gg)
  
}
```


```{r}
# Structure functions:

GetMFE <- function(L_sequence, R_sequence) {
  
  input <- paste0(L_sequence, "\n", R_sequence)
  rnaduplex <- system("RNAduplex --noLP", input = input, intern = TRUE)
  
  rnaduplex <- gsub("\\s+", "_", rnaduplex)
  mfe <- sapply(strsplit(rnaduplex, "_"), "[", 5) # Need to fix for positives < 10 ? as they have an extra space
  mfe <- as.numeric(gsub("\\(|\\)", "", mfe))
  
  # Get dot.bracket
  db <- sapply(strsplit(rnaduplex, "_"), "[", 1)
  
  # Get from/to positions
  l <- sapply(strsplit(rnaduplex, "_"), "[", 2)
  r <- sapply(strsplit(rnaduplex, "_"), "[", 4)
  l_from <- as.numeric(sapply(strsplit(l, ","), "[", 1))
  l_to <- as.numeric(sapply(strsplit(l, ","), "[", 2))
  r_from <- as.numeric(sapply(strsplit(r, ","), "[", 1))
  r_to <- as.numeric(sapply(strsplit(r, ","), "[", 2))
  
  l_width <- nchar(L_sequence)
  r_width <- nchar(R_sequence)
  
  # Pad db with . according to from/to
  l_db <- sapply(strsplit(db, "&"), "[", 1)
  r_db <- sapply(strsplit(db, "&"), "[", 2)
  l_db <- unlist(strsplit(l_db, "")) # convert string to vector of characters
  r_db <- unlist(strsplit(r_db, ""))
  
  l_db <- c(rep(".", l_from - 1L), l_db, rep(".", l_width - l_to))

  r_db <- c(rep(".", r_from - 1L), r_db, rep(".", r_width - r_to))
  
  stopifnot(all(length(l_db == l_width), length(r_db) == r_width))
  db <- paste0(c(l_db, "&", r_db), collapse = "")

  #if(mfe > 0) mfe <- 0
  
 return(list(mfe = mfe, db = db))
  
}

# find position of the first "(" and ")" in L and R respectively
# find position of the last "(" and ")" in L and R respectively

# trim external dots..to get all paired (find pos of first and last ())
# trim same positions from seq

get_paired_regions <- function(id, L_db, L_sequence, R_db, R_sequence) {
  
  l_brackets <- str_locate_all(L_db, pattern = "\\(")[[1]][,1]
  l_start <- l_brackets[1]
  l_end <- l_brackets[length(l_brackets)]
  l_trimmed_db <- str_sub(L_db, l_start, l_end)
  l_trimmed_seq <- str_sub(L_sequence, l_start, l_end)

  r_brackets <- str_locate_all(R_db, pattern = "\\)")[[1]][,1]
  r_start <- r_brackets[1]
  r_end <- r_brackets[length(r_brackets)]
  r_trimmed_db <- str_sub(R_db, r_start, r_end)
  r_trimmed_seq <- str_sub(R_sequence, r_start, r_end)

  stopifnot(length(l_brackets) == length(r_brackets)) # check L and R arms of the duplex have an equal number of bp

  paired_residues_count <- length(l_brackets)
  l_paired_residues <- paste(str_sub(L_sequence, start = l_brackets, end = l_brackets), collapse = '')
  r_paired_residues <- paste(str_sub(R_sequence, start = r_brackets, end = r_brackets), collapse = '')

  results.ls <- list(id = id, l_start =l_start, l_end = l_end, l_trimmed_db=l_trimmed_db, l_trimmed_seq=l_trimmed_seq,
                   r_start=r_start, r_end=r_end, r_trimmed_db=r_trimmed_db, r_trimmed_seq=r_trimmed_seq,
                   paired_residues_count = paired_residues_count,
                   l_paired_residues=l_paired_residues, r_paired_residues=r_paired_residues)
  
  return(results.ls)

}


#### functions to get max paired
get_stem_segment_length <- function(input.ls) {
  
  data <- input.ls$data
  
  if (is.null(data)) {
    input.ls <- input.ls
    stem_len = 0
  } else {
    
    if (!FALSE %in% unique(data$l_bracket_pos == data$r_bracket_pos)) {
      stem_len <- nrow(data)
      data <- data
      input.ls <- list(data=data, stem_len=stem_len)
      
    } else {
      data <- data %>%
        mutate(l_bracket_pos = l_bracket_pos - as.numeric(data[1,1]),
                r_bracket_pos = r_bracket_pos - as.numeric(data[1,2]))
    
      stem.df <- data %>% dplyr::filter(l_bracket_pos == r_bracket_pos)
      stem_len <- nrow(stem.df) 
      
      rest.df <- anti_join(data, stem.df, by = "pos")
    
      if (nrow(rest.df) == 0) {
        input.ls <- list(data = stem.df, stem_len=stem_len)
      } else {
        
        input.ls <- list(data = rest.df, stem_len=stem_len)
      }
    }
      
  }
  write(stem_len, file = "stem_lengths.txt", sep="\n", append = TRUE)
  
  return(input.ls)
}



get_max_stem_length <- function(input.ls, i=30) {
  
  if (i == 0) {
    stem_len <- get_stem_segment_length(input.ls)$stem_len
    
    input.ls <- list(input.ls$data, stem_len)
    
    return(input.ls)
    
    } else {
                  
      input.ls <- get_stem_segment_length(input.ls)
      i <- i - 1
      get_max_stem_length(input.ls, i)
    }
  
  stem_lengths <- read.csv("stem_lengths.txt", sep="\t")
  col <- colnames(stem_lengths)
  max <- max(stem_lengths[,col])
  
  return(max)
 
}

get_max_paired <- function(dataframe, id) {
  
  data <- dataframe[dataframe$id == id,]
  
  L_db <- data$L_duplex_db
  R_db <- data$R_duplex_db
  
  rev_R_db <- stringi::stri_reverse(R_db)
  l_bracket <- str_locate_all(L_db, pattern = "\\(")[[1]][,1]
  r_bracket <- str_locate_all(rev_R_db, pattern = "\\)")[[1]][,1]
  
  paired.dt <- as.data.table(list(l_bracket_pos = l_bracket, r_bracket_pos=r_bracket))
  paired.dt$pos <- row.names(paired.dt)
  
  input.ls <- list(data = paired.dt, stem_len = 1)
  
  max <- get_max_stem_length(input.ls)
  max.dt <- as.data.table(list(id = id, max_paired = max))
  
  file.remove("stem_lengths.txt")
  
  return(max.dt)
  
}
```


# Get genomic coordinates and annotate STAU1 hybrids

#### This step was necessary because a bug in Tosca is not getting genomic coordinates or annotating most hybrids. The bug can be fixed if you reformat names of hybrids seqnames so they match the annotation gtf:
```{r eval=FALSE, echo=TRUE}
hybrids.dt[, c("L_seqnames") := tstrsplit(L_seqnames, "::", fixed=TRUE)[1]]
hybrids.dt[, c("R_seqnames") := tstrsplit(L_seqnames, "::", fixed=TRUE)[1]]
```
#### before calling the 'convert_coordinates' function.


```{r, warning = FALSE}
genes.gr <- rtracklayer::import.gff2("/Users/iosubi/Documents/projects/computational_hiCLIP/files/human_GencodeV33.gtf.gz")
regions.gr <- import.gff2("~/Documents/Genomes/human/regions.gtf.gz")
hybrids.dt <- fread("/Users/iosubi/Documents/projects/computational_hiCLIP/files/all.clusters.tsv.gz")

hybrids.dt <- hybrids.dt[grep("^rRNA", L_seqnames, invert = TRUE)] # Remove rRNA
hybrids.dt <- hybrids.dt[grep("Mt", L_seqnames, invert = TRUE)]

# Reformat names of hybrids seqnames so they match the annotation gtf
hybrids.dt[, c("L_seqnames") := tstrsplit(L_seqnames, "::", fixed=TRUE)[1]]
hybrids.dt[, c("R_seqnames") := tstrsplit(L_seqnames, "::", fixed=TRUE)[1]]

# Convert to genomic coordinates
hybrids.dt <- convert_coordinates(hybrids.dt, genes.gr)
#fwrite(hybrids.dt, "clusters.gc.tsv.gz", sep = "\t")

# Annotate hybrids
hybrids.dt <- annotate_hybrids(hybrids.dt, regions.gr)
#fwrite(hybrids.dt, "annotated_clusters.gc.tsv.gz", sep = "\t")

hybrids.dt$Experiment <- "STAU1 - No linker"
```

# PARIS data: Get genomic coordinates and annotate hybrids

```{r, warning = FALSE}
paris.dt <- fread("/Users/iosubi/Documents/projects/computational_hiCLIP/files/paris/all.clusters.tsv.gz")
paris.dt <- paris.dt[grep("^rRNA", L_seqnames, invert = TRUE)] # Remove rRNA
paris.dt <- paris.dt[grep("Mt", L_seqnames, invert = TRUE)]

paris.dt[, c("L_seqnames") := tstrsplit(L_seqnames, "::", fixed=TRUE)[1]]
paris.dt[, c("R_seqnames") := tstrsplit(L_seqnames, "::", fixed=TRUE)[1]]

# Convert to genomic coordinates
paris.dt <- convert_coordinates(paris.dt, genes.gr)
#fwrite(paris.dt, "paris_clusters.gc.tsv.gz", sep = "\t")

# Annotate hybrids
paris.dt <- annotate_hybrids(paris.dt, regions.gr)
#fwrite(paris.dt, "paris_annotated_clusters.gc.tsv.gz", sep = "\t")
paris.dt$Experiment <- "PARIS"
```


#### Merge experiments

```{r}
all_exp.df <- as.data.frame(rbind(hybrids.dt, paris.dt))
# intramolecular only
all_exp.df <- all_exp.df %>% dplyr::filter(L_gene_name == R_gene_name)
```


# Transcript biotypes
```{r}
biotype_counts.df <- all_exp.df %>%
  select(name, Experiment, L_biotype, R_biotype) %>%
  pivot_longer(c(L_biotype, R_biotype), names_to = "arm", values_to = "biotype")

biotype_counts.df <- biotype_counts.df %>%
  group_by(Experiment, biotype) %>%
  mutate(biotype = case_when(str_detect(biotype, "lncRNA") ~ "lncRNA",
                             TRUE ~ biotype)) %>%
  summarise(counts = n()) %>%
  arrange(desc(biotype)) %>%
  mutate(percentage = counts*100 / sum(counts)) %>%
  ungroup()
biotype_counts.df <- biotype_counts.df %>%
  mutate(RNA_biotype = case_when((percentage < 0.05) ~ "Other",
                                 (percentage >= 0.05) ~ biotype))

biotype_counts.df <- biotype_counts.df %>%
  group_by(Experiment, RNA_biotype) %>%
  mutate(biotype_percentage = sum(percentage)) %>%
  ungroup()

```

```{r echo=FALSE}
biotype_counts.df
```

```{r}
# Simple bar-chart

ggplot(distinct(biotype_counts.df, across(c(Experiment, RNA_biotype, biotype_percentage))),
       aes(RNA_biotype, biotype_percentage)) + 
  geom_bar(position="dodge", stat = "identity", aes(fill = Experiment), alpha = 0.8) +
  scale_fill_brewer(palette="Dark2") +
  #scale_fill_paletteer_d("rcartocolor::Earth", direction = -1) +
  theme(legend.position="right", legend.direction="vertical",
        legend.title = element_blank()) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ylab("Percentage") +
  xlab("RNA biotype")
```

# mRNA transcript regions

```{r}
all_exp_mRNA.df <- all_exp.df %>% dplyr::filter(R_biotype  == "mRNA" & L_biotype == "mRNA")

all_exp_mRNA.df <- all_exp_mRNA.df %>%
  mutate(hybrid_type = case_when((L_region == "UTR3" & R_region == "UTR3") ~ "3'UTR-3'UTR",
                                 (L_region == "CDS" & R_region == "CDS") ~ "CDS-CDS",
                                 (L_region == "UTR3" & R_region == "CDS") ~ "3'UTR-CDS",
                                 (L_region == "CDS" & R_region == "UTR3") ~ "3'UTR-CDS",
                                 (L_region == "UTR5" & R_region == "UTR5") ~ "5'UTR-5'UTR",
                                 (L_region == "UTR3" & R_region == "UTR5") ~ "3'UTR-5'UTR",
                                 (L_region == "UTR5" & R_region == "UTR3") ~ "3'UTR-5'UTR",
                                 (L_region == "UTR5" & R_region == "CDS") ~ "5'UTR-CDS",
                                 (L_region == "CDS" & R_region == "UTR5") ~ "5'UTR-CDS"))


plot_stacked_barchart(all_exp_mRNA.df, hybrid_type, Experiment)

```


# 3'UTR hybrids

```{r}
# Get hybrid spans
# Get hybrid arm widths

threeutr.df <- all_exp_mRNA.df %>%
  dplyr::filter(L_region == "UTR3" & R_region == "UTR3")

threeutr.df <-threeutr.df %>%
  rowwise() %>%
  mutate(hybrid_span = R_start - L_end + 1, 
         L_arm_width = L_end - L_start + 1,
         R_arm_width = R_end - R_start + 1)
```

## Hybrid span

```{r}
ggplot(threeutr.df, aes(x = hybrid_span, color = Experiment)) + 
  geom_density(alpha = 0.8) +
  scale_colour_brewer(palette="Dark2") +
  scale_x_log10() + annotation_logticks() +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ylab("Density") +
  xlab("Hybrid span")
```

## Arm width

```{r}
# convert to long format
threeutr_long.df <- threeutr.df %>%
  pivot_longer(c(R_arm_width, L_arm_width), names_to = "arm", values_to = "arm_width")

ggplot(threeutr_long.df, aes(x = arm, y = arm_width, fill = Experiment)) +
  geom_boxplot(alpha = 0.7) +
  theme(axis.text.x = element_text(angle = 90)) +
  scale_fill_brewer(palette="Dark2") +
  theme(legend.position="right", legend.direction="vertical") +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ylab("Hybrid arm width") +
  xlab("Arm")
```

# Binding energy

```{r}
# Load data that contains MFE information for the hybrids

hybrids_mfe.dt <- fread("/Users/iosubi/Documents/projects/computational_hiCLIP/files/all.mfe.clusters.tsv.gz")

hybrids_mfe.dt <- hybrids_mfe.dt[grep("^rRNA", L_seqnames, invert = TRUE)] # Remove rRNA
hybrids_mfe.dt <- hybrids_mfe.dt[grep("Mt", L_seqnames, invert = TRUE)]

hybrids_mfe.dt <- hybrids_mfe.dt %>%
  select(-umi) %>%
  mutate(Experiment = "STAU1 - No linker")

paris_mfe.dt <- fread("/Users/iosubi/Documents/projects/computational_hiCLIP/files/paris/all.mfe.clusters.tsv.gz")
paris_mfe.dt$Experiment = "PARIS"

paris_mfe.dt <- paris_mfe.dt[grep("^rRNA", L_seqnames, invert = TRUE)] # Remove rRNA
paris_mfe.dt <- paris_mfe.dt[grep("Mt", L_seqnames, invert = TRUE)]

all_exp_mfe.dt <- rbind(hybrids_mfe.dt, paris_mfe.dt)
```



## 3'UTR hybrids MFE

```{r, warning = FALSE}
# Annotate hybrids
all_exp_mfe.dt[, c("L_seqnames") := tstrsplit(L_seqnames, "::", fixed=TRUE)[1]]
all_exp_mfe.dt[, c("R_seqnames") := tstrsplit(L_seqnames, "::", fixed=TRUE)[1]]

all_exp_mfe.dt <- convert_coordinates(all_exp_mfe.dt, genes.gr)
all_exp_mfe.dt <- annotate_hybrids(all_exp_mfe.dt, regions.gr)

mfe_3utr.dt <- all_exp_mfe.dt %>%
  dplyr::filter(L_region == "UTR3" & R_region == "UTR3")

ggplot(mfe_3utr.dt, aes(x = mfe, color = Experiment)) + 
  geom_density(alpha = 0.8) +
  scale_colour_brewer(palette="Dark2") +
  facet_grid(cols = vars(Experiment)) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ylab("Density") +
  xlab("MFE (kcal/mol)") +
  ggtitle("3'UTR hybrids")

```

## 3'UTR clusters MFE


```{r, warning = FALSE}
# obtain minimum mfe per cluster per experiment
mfe_3utr.dt <- mfe_3utr.dt %>%
  dplyr::filter(str_detect(cluster, "C")) %>%
  group_by(Experiment, cluster, L_gene_id, R_gene_id) %>%  #had to group also by gene id because redundant cluster names in PARIS data
  mutate(min_mfe = min(mfe)) %>%
  ungroup() %>%
  select(Experiment, cluster, min_mfe) %>%
  distinct()
```

```{r, warning = FALSE}
# fuse min_mfe information to the clusters
# threeutr.df <- left_join(threeutr.df, mfe_3utr.dt, by = c("name" = "cluster"))
ggplot(mfe_3utr.dt, aes(x = min_mfe, color = Experiment)) + 
  geom_density(alpha = 0.8) +
  scale_colour_brewer(palette="Dark2") +
  facet_grid(cols = vars(Experiment)) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank()) +
  ylab("Density") +
  xlab("MFE (kcal/mol)") +
  ggtitle("3'UTR clusters")

```


# Hybrid clusters with mega-long arms

```{r}
long_arm_clusters.df <- threeutr.df %>% dplyr::filter(L_arm_width > 90 | R_arm_width > 90)
long_arm_hybrids.dt <- semi_join(hybrids_mfe.dt, long_arm_clusters.df, by = c("cluster" = "name"))

```

# Binding energy (updated)


```{r}
hybrids_mfe.dt <- hybrids_mfe.dt %>% dplyr::filter(str_detect(cluster, "C")) #keep hybrids within clusters

# for each cluster, select the hybrid corresponding to min MFE, only these hybrids will be analysed further
threeutr.dt <- hybrids_mfe.dt 

threeutr.dt <- threeutr.dt %>%
  dplyr::filter(str_detect(cluster, "C")) %>%
  group_by(cluster) %>%  #had to group also by gene id because redundant cluster names in PARIS data
  dplyr::slice(which.min(mfe)) %>%
  ungroup()

```

```{r}
# get 3'UTR cluster names from the clusters table
threeutr_clusters.dt <- hybrids.dt %>%
  dplyr::filter(L_region == "UTR3" & R_region == "UTR3") 

threeutr_clusters.ls <- unique(threeutr_clusters.dt$name)

# filter hybrids table by 3'UTR cluster names
threeutr.dt <- threeutr.dt[threeutr.dt$cluster %in% threeutr_clusters.ls,]
```

## Get dot bracket structure

```{r}
threeutr.dt <- threeutr.dt %>%
  rowwise() %>%
  mutate(binding = list(GetMFE(L_sequence, R_sequence))) %>%
  mutate(db = binding$db, mfe = binding$mfe) %>%
  ungroup() %>%
  dplyr::select(-binding)

```

## Get paired positions, db, sequences, total sum of paired residues within a duplex

```{r}
# separate L_db and R_db
threeutr.dt <- threeutr.dt %>%
  separate(db, into = c("L_db", "R_db"), sep = "&", remove = FALSE)
```


```{r}
threeutr.dt <- threeutr.dt %>%
  rowwise() %>%
  mutate(paired_regions = list(get_paired_regions(id, L_db, L_sequence, R_db, R_sequence))) %>%
  mutate(L_duplex_start= paired_regions$l_start, 
         L_duplex_end = paired_regions$l_end,
         L_duplex_db =  paired_regions$l_trimmed_db,
         L_duplex_sequence =  paired_regions$l_trimmed_seq,
         R_duplex_start= paired_regions$r_start, 
         R_duplex_end = paired_regions$r_end,
         R_duplex_db =  paired_regions$r_trimmed_db,
         R_duplex_seq =  paired_regions$r_trimmed_seq,
         total_paired = paired_regions$paired_residues_count,
         l_paired_residues = paired_regions$l_paired_residues,
         r_paired_residues = paired_regions$r_paired_residues) %>%
  ungroup() %>%
  dplyr::select(-paired_regions)
```

## Get the maximum length of uninterrupted duplex segment within each hybrid

```{r}
max_paired.dt <- rbindlist(lapply(threeutr.dt$id, get_max_paired, data=threeutr.dt))
threeutr.dt <- left_join(threeutr.dt, max_paired.dt, by = "id")
```

## Count the maximum number of internal loops/bulges per duplex

```{r}
threeutr.dt <- threeutr.dt %>%
  rowwise() %>%
  mutate(L_iloop_count = length(str_locate_all(L_duplex_db, pattern = fixed("(."))[[1]][,1]),
         R_iloop_count = length(str_locate_all(R_duplex_db, pattern = fixed(".)"))[[1]][,1])) %>%
  mutate(iloop_counts = max(L_iloop_count, R_iloop_count)) %>%
  ungroup()
```


```{r}
# fwrite(file = "threeutr.db.tsv", threeutr.dt, sep="\t")
```

## Analyse/visualise duplex features

```{r}
id_features.df <- threeutr.dt %>%
  dplyr::select(id, total_paired, max_paired, iloop_counts) %>%
  ungroup()
```


```{r}
quantile(id_features.df$total_paired, 0.9)
quantile(id_features.df$total_paired, 0.1)
median(id_features.df$total_paired)
```

```{r}
quantile(id_features.df$max_paired, 0.9)
quantile(id_features.df$max_paired, 0.1)
median(id_features.df$max_paired)
```

```{r}
quantile(id_features.df$iloop_counts, 0.9)
quantile(id_features.df$iloop_counts, 0.1)
median(id_features.df$iloop_counts)
```



```{r}
# convert to long format
id_features.df <- rowid_to_column(id_features.df)
id_features.df <- id_features.df %>% 
  gather("feature", "counts", total_paired:iloop_counts)

id_features_summed.gg <- ggplot(id_features.df, aes(x = factor(feature, levels = c("total_paired", "max_paired","iloop_counts")), y = counts, fill = feature)) +
  geom_boxplot() +
  ylim(0, 72) +
  theme(axis.text.x = element_text(angle = 90)) +
  scale_fill_brewer(palette="Blues") +
  theme(legend.position="right", legend.direction="vertical") +
  ggtitle("STAU1 hybrids")

id_features_summed.gg
ggsave(paste0("linker_boxplot.pdf"), id_features_summed.gg, height = 7, width = 7, dpi = 300)

```

# Sequence content

## Nucleotide frequency - only the paired residues

```{r}
paired_nuc_freq.df <- get_nucleotide_frequency(threeutr.dt)
paired_nuc_freq.gg <- plot_nucleotide_frequency(paired_nuc_freq.df)
paired_nuc_freq.gg
ggsave(paste0("nolinker_nuc_freq.pdf"), paired_nuc_freq.gg, height = 11, width = 7, dpi = 300)
```


```{r}
# df <- df %>%
#   rowwise() %>%
#   mutate(L_iloop_starts = str_locate_all(L_duplex_db, pattern = fixed("(."))[[1]][,1],
#          L_iloop_ends = str_locate_all(L_duplex_db, pattern = fixed(".("))[[1]][,1],
#          R_rev = stringi::stri_reverse(R_duplex_db)) %>%
#   mutate(R_iloop_starts = str_locate_all(R_rev, pattern = fixed("(."))[[1]][,1],
#          R_iloop_ends = str_locate_all(R_rev, pattern = fixed(".("))[[1]][,1]) %>%
#   
```




