---
title: "STAU1 RNA metabolism analysis"
author: "A. M. Chakrabarti"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  html_notebook: 
    code_folding: hide
    toc: yes
    toc_depth: 2
    toc_float: yes
    number_sections: yes
    theme: flatly
    highlight: haddock
    fig_align: center
---

```{r, echo = FALSE}
library(primavera)
library(data.table)
library(ggplot2)
library(cowplot)
library(scales)
library(ggthemes)
library(UpSetR)
```

# Load data

```{r}
linker.dt <- fread("~/projects/comp_hiclip/linker/linker.clusters.mfe.tsv.gz")
linker.clusters <- collapse_clusters(linker.dt[L_seqnames == R_seqnames][cluster_hybrid_count >= 2], mode = "wide")

nolinker.dt <- fread("~/projects/comp_hiclip/no_linker/atlas/all.hybrids.tsv.gz")
nolinker.dt[, sample := tstrsplit(sample, "\\.")[[1]]]
nolinker.clusters <- collapse_clusters(nolinker.dt[L_seqnames == R_seqnames][cluster_hybrid_count >= 2], mode = "wide")

linker.clusters <- primavera::reorient_hybrids(linker.clusters)
nolinker.clusters <- primavera::reorient_hybrids(nolinker.clusters)
```

```{r}
atlas.hybrids.dt <- linker.dt[, total_count := .N, by = .(L_seqnames, R_seqnames)]
atlas.hybrids.dt <- atlas.hybrids.dt[L_seqnames == R_seqnames]
atlas.hybrids.dt <- atlas.hybrids.dt[!(L_seqnames == "rRNA_45S" & R_seqnames == "rRNA_45S")]
atlas.hybrids.dt <- atlas.hybrids.dt[!(L_seqnames == "rDNA" & R_seqnames == "rDNA")]
atlas.hybrids.dt <- atlas.hybrids.dt[!(L_seqnames == "rRNA_5S" & R_seqnames == "rRNA_5S")]
atlas.hybrids.dt <- atlas.hybrids.dt[!(L_seqnames == "rRNA5S" & R_seqnames == "rRNA5S")]
atlas.hybrids.dt <- atlas.hybrids.dt[!grepl("tRNA", L_seqnames)]
atlas.hybrids.dt <- atlas.hybrids.dt[!grepl("tRNA", R_seqnames)]

linker.list <- split(atlas.hybrids.dt, by = c("L_seqnames", "R_seqnames"))
linker.clusters.list <- parallel::mclapply(linker.list, cluster_hybrids, percent_overlap = 0.5, mc.cores = 4)
linker.clusters.dt <- rbindlist(linker.clusters.list, use.names = TRUE, fill = TRUE)

linker.clusters.dt[L_seqnames == R_seqnames][!L_seqnames %in% c("tRNA", "rDNA", "rRNA_5S")][grep("C", cluster), .N, by = .(cluster, L_seqnames, R_seqnames)]
nolinker.dt[L_seqnames == R_seqnames][grep("C", cluster), .N, by = .(cluster, L_seqnames, R_seqnames)]

dt <- data.table(exp = c("Linker", "No Linker"),
                 counts = c(nrow(linker.clusters.dt[L_seqnames == R_seqnames][!L_seqnames %in% c("tRNA", "rDNA", "rRNA_5S")][grep("C", cluster), .N, by = .(cluster, L_seqnames, R_seqnames)]),
                            nrow(nolinker.dt[L_seqnames == R_seqnames][grep("C", cluster), .N, by = .(cluster, L_seqnames, R_seqnames)])))

p <- ggplot(dt, aes(x = exp, y = counts)) +
  geom_col(width = 0.5) +
  scale_fill_tableau() +
  theme_minimal_grid() + theme(legend.position = "top") +
  labs(x = "",
       y = "Number of clusters",
       fill = "")

p
```


# Get overlaps

```{r}
bedpe.colnames <- c("L_seqnames", "L_start", "L_end", "R_seqnames", "R_start", "R_end", "name", "count", "L_strand", "R_strand")
linker.bedpe.dt <- linker.duplexes.dt[, ..bedpe.colnames]
linker.bedpe.dt[, `:=`(
  L_start = L_start - 1,
  R_start = R_start - 1
)]
nolinker.bedpe.dt <- nolinker.clusters[, ..bedpe.colnames]
nolinker.bedpe.dt[, `:=`(
  L_start = L_start - 1,
  R_start = R_start - 1
)]

linker.bedpe <- tempfile(tmpdir = getwd(), fileext = ".bedpe")
nolinker.bedpe <- tempfile(tmpdir = getwd(), fileext = ".bedpe")
ol <- tempfile(tmpdir = getwd(), fileext = ".bedpe")

fwrite(linker.bedpe.dt, file = linker.bedpe, sep = "\t", col.names = FALSE)
fwrite(nolinker.bedpe.dt, file = nolinker.bedpe, sep = "\t", col.names = FALSE)

cmd <- paste("bedtools pairtopair -rdn -a", linker.bedpe, "-b", nolinker.bedpe, ">", ol)
system(cmd)
bedpe.dt <- fread(ol, col.names = c(paste0(bedpe.colnames, ".x"), paste0(bedpe.colnames, ".y")))

file.remove(linker.bedpe)
file.remove(nolinker.bedpe)
file.remove(ol)
```

```{r}
library(eulerr)

fit <- euler(c("Linker" = nrow(linker.duplexes.dt),
               "No Linker" = nrow(nolinker.clusters),
               "Linker&No Linker" = nrow(bedpe.dt)))

plot(fit)

```

```{r}
upset.input <- c("Linker" = nrow(linker.duplexes.dt),
               "No Linker" = nrow(nolinker.clusters),
               "Linker&No Linker" = nrow(bedpe.dt))

upset(fromExpression(upset.input), 
      order.by = "freq",
      mainbar.y.label = "Number of duplexes")
```

```{r}
bam.files <- c("~/projects/comp_hiclip/linker/LigPlusHighAligned.sortedByCoord.out.bam",
               "~/projects/comp_hiclip/linker/LigPlusLowAligned.sortedByCoord.out.bam")

bam.list <- lapply(bam.files, convert_linkerbam_to_hybrids)
hybrids.dt <- rbindlist(bam.list)
hybrids.dt$exp <- rep(c("LigPlusHigh", "LigPlusLow"), S4Vectors::elementNROWS(bam.list))


```


```{r}
silico.dt <- rbindlist(list(fread("~/projects/comp_hiclip/results_silico/hybrids/LigPlusHighSilico.hybrids.tsv.gz")[, exp := "high RNase"],
                              fread("~/projects/comp_hiclip/results_silico/hybrids/LigPlusLowSilico.hybrids.tsv.gz")[, exp := "low RNase"]))

hybrids.dt[!name %in% silico.dt$name, silico := FALSE]
hybrids.dt[is.na(silico), silico := TRUE]

prop.table(table(hybrids.dt$silico))

ggplot(hybrids.dt, aes(x = exp, fill = silico)) +
  geom_bar() +
  scale_fill_tableau() +
  theme_minimal_grid() + theme(legend.position = "bottom") +
  geom_vline(xintercept = 60, linetype = "dotted", colour = "red") +
  labs(x = "",
       y = "Number of hybrids",
       fill = "Recovered from in silico read") +
  coord_flip()

```

```{r}
ggplot(hybrids.dt, aes(x = seq_width, colour = silico, linetype = exp)) +
  geom_density() +
  scale_color_tableau() +
  theme_minimal_grid() + theme(legend.position = "bottom") +
  geom_vline(xintercept = 60, linetype = "dotted", colour = "red") +
  labs(x = "In silico read length",
       y = "Density",
       colour = "Recovered from in silico read",
       linetype = "")
```

```{r}
library(rtracklayer)
genes.gr <- import.gff2("~/Dropbox (The Francis Crick)/comp_hiclip/ref/GRCh38.gencode_v33.tx.gtf.gz")
regions.gr <- import.gff2("~/Dropbox (The Francis Crick)/comp_hiclip/ref/regions.gtf.gz")

n <- names(hybrids.dt)
n[grep("\\.x$", n)] <- paste0("L_", gsub("\\.x$", "", n[grep("\\.x$", n)]))
n[grep("\\.y$", n)] <- paste0("R_", gsub("\\.y$", "", n[grep("\\.y$", n)]))
n <- gsub("_s_", "_", n)
n <- gsub("_subject", "_seqnames", n)
setnames(hybrids.dt, n)
hybrids.dt[, `:=`(L_strand = "+", R_strand = "+")]

linker.dt <- convert_coordinates(hybrids.dt = hybrids.dt, genes.gr = genes.gr)
linker.dt <- annotate_hybrids(hybrids.dt = linker.dt, regions.gr = regions.gr)
linker.dt[, type := ifelse(L_seqnames == R_seqnames, "intragenic", "intergenic")]

ggplot(linker.dt[, .N, by = .(exp, silico, type)], aes(x = silico, y = N, fill = type)) +
  geom_col(position = "fill") +
  scale_y_continuous(label = percent) +
  scale_fill_tableau(palette = "Seattle Grays") +
  facet_grid(exp ~ .) +
  theme_minimal_grid() + theme(legend.position = "top") +
  labs(x = "Recovered from in silico read",
       y = "Percentage of hybrids",
       fill = "") +
  coord_flip()
```

```{r}
arms.dt <- rbindlist(list(linker.dt[, .(exp, silico, type, L_region)], linker.dt[, .(exp, silico, type, R_region)]), use.names = FALSE)
arms.dt$L_region <- factor(arms.dt$L_region, levels = c("rRNA", "tRNA", "UTR5", "CDS", "intron", "UTR3", "ncRNA"))
p <- ggplot(arms.dt[, .N, by = .(L_region, exp, type, silico)], aes(x = silico, y = N, fill = L_region)) +
  geom_col(position = "fill") +
  scale_y_continuous(label = percent) +
  scale_fill_tableau(palette = "Miller Stone") +
  facet_grid(exp ~ type) +
  theme_minimal_grid() + theme(legend.position = "top") +
  labs(x = "Recovered from in silico read",
       y = "Percentage of hybrids",
       fill = "") +
  coord_flip()

p
```

