---
title: "Figure 2"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(pheatmap)
library(viridis)
library(RColorBrewer)
library(tictoc)
library(primavera)
library(rtracklayer)
library(data.table)
library(ggplot2)
library(scales)
library(ggthemes)
library(cowplot)
library(UpSetR)
library(tidyverse)
library(ggrepel)
```

```{r}
plot.path <- "~/Dropbox (The Francis Crick)/comp_hiclip/plots/figure_1"
if(!dir.exists(plot.path)) dir.create(plot.path)

ht_colours <- c("Linker" = "#8175aa", "Direct" = "#6fb899", "Short-range" = "#31a1b3")
```

## Load data

```{r}
# Linker
# linker.dt <- fread("/Volumes/lab-luscomben/home/shared/projects/ira-nobby/comp_hiclip/results_linker/linker.clusters.mfe.tsv.gz")
linker.dt <- rbindlist(list(
  fread("/Volumes/lab-luscomben/home/shared/projects/ira-nobby/comp_hiclip/results_linker/lph.hybrids.dedup.tsv.gz")[, sample := "stau1_high"],
  fread("/Volumes/lab-luscomben/home/shared/projects/ira-nobby/comp_hiclip/results_linker/lpl.hybrids.dedup.tsv.gz")[, sample := "stau1_low"]),
                       use.names = TRUE)
# Add calculations
linker.dt[, `:=` (L_end = L_start + L_width - 1,
                  R_end = R_start + R_width - 1)]
linker.dt[, type := ifelse(L_seqnames == R_seqnames, "intragenic", "intergenic")]
linker.dt[, orientation := "linker"]

genes.gr <- import.gff2("~/Dropbox (The Francis Crick)/comp_hiclip/ref/GRCh38.gencode_v33.tx.gtf.gz")
regions.gr <- import.gff2("~/Dropbox (The Francis Crick)/comp_hiclip/ref/regions.gtf.gz")

linker.dt <- convert_coordinates(hybrids.dt = linker.dt, genes.gr = genes.gr)
linker.dt <- annotate_hybrids(hybrids.dt = linker.dt, regions.gr = regions.gr)

# No linker
nolinker.dt <- fread("/Volumes/lab-luscomben/home/shared/projects/ira-nobby/comp_hiclip/results_nolinker/atlas/all.hybrids.tsv.gz")
nolinker.dt[, sample := tstrsplit(sample, "\\.")[[1]]]

hybrids.dt <- rbindlist(list(nolinker.dt, linker.dt),
                        use.names = TRUE,
                        fill = TRUE)
hybrids.dt[, exp := ifelse(orientation == "linker", "Linker", "Direct")]
hybrids.dt[sample == "stau1_high", sample := "High RNase"]
hybrids.dt[sample == "stau1_low", sample := "Low RNase"]

hybrids.dt[, type := ifelse(type == "intragenic", "intra-transcript", "inter-transcript")]

```

# Figure 2 A - Counts

```{r}
p <- ggplot(hybrids.dt[, .N, by = .(exp, sample)], aes(x = exp, y = N, fill = sample)) +
  geom_col(width = 0.5) +
  coord_flip() + scale_x_discrete(limits = c("Direct", "Linker")) +
  scale_fill_tableau() +
  theme_minimal_grid() + theme(legend.position = "top") +
  labs(x = "",
       y = "Number of hybrids",
       fill = "")

ggsave(p, filename = file.path(plot.path, "linker_nolinker_counts.pdf"), width = 7, height = 4)

p
```

There are `r nrow(nolinker.dt)/nrow(linker.dt)` times as many hybrids recovered without a linker

# Figure 2 B - Intra-inter

```{r}

p <- hybrids.dt %>% 
  group_by(exp, type) %>% 
  summarise(reads = n()) %>%
  mutate(percentage = reads/sum(reads)) %>%
  ggplot(aes(x = exp, y = percentage, fill = type)) +
  geom_col(width = 0.5, position = "fill") +
  geom_text(aes(label = paste0(round(percentage * 100,1),"%")), 
            position = position_stack(vjust=0.5),
            colour = "white",
            size = 5) +
  coord_flip() + scale_x_discrete(limits = c("Direct", "Linker")) +
  scale_y_continuous(label = percent) +
  scale_fill_tableau(palette = "Seattle Grays", guide = guide_legend(reverse = TRUE)) +
  theme_minimal_grid() + theme(legend.position = "top") +
  labs(x = "",
       y = "Percentage of hybrids",
       fill = "")

ggsave(p, filename = file.path(plot.path, "linker_nolinker_interintra.pdf"), width = 7, height = 4)

p
```


# Figure 2 C - Regions

```{r}
arms.dt <- rbindlist(list(hybrids.dt[, .(exp, L_region)], hybrids.dt[, .(exp, R_region)]), use.names = FALSE)
arms.dt$L_region <- factor(arms.dt$L_region, levels = rev(c("rRNA", "tRNA", "UTR5", "CDS", "intron", "UTR3", "ncRNA")))

# p <- arms.dt %>%
#   group_by(L_region, exp) %>%
#   summarise(arms = n()) %>%
#   group_by(exp) %>%
#   mutate(percentage = arms/sum(arms)) %>%
#   arrange(desc(L_region)) %>%
#   mutate(pos = cumsum(percentage) - percentage/2) %>%
#   ggplot(aes(x = exp, y = percentage, fill = L_region)) +
#   geom_col(width = 0.5, position = "fill") +
#   geom_text_repel(aes(y = pos, label = paste0(round(percentage * 100,1),"%")),
#           size = 5,
#           nudge_x = 0.4,
#           segment.size = 0.5, 
#           segment.color = '#515A5A') +
#   coord_flip() +
#   scale_x_discrete(limits = c("Direct", "Linker")) +
#   scale_y_continuous(label = percent) +
#   scale_fill_tableau(palette = "Miller Stone", direction = -1) +
#   theme_minimal_grid() + theme(legend.position = "top") +
#   labs(x = "",
#        y = "Percentage of hybrids",
#        fill = "")

p <- arms.dt %>%
  group_by(L_region, exp) %>%
  summarise(arms = n()) %>%
  group_by(exp) %>%
  mutate(percentage = arms/sum(arms)) %>%
  ggplot(aes(x = exp, y = percentage, fill = L_region)) +
  geom_col(width = 0.5, position = "fill") +
  geom_text(aes(label = ifelse(percentage > 0.05, paste0(round(percentage * 100, 1), "%"), "")), 
            position = position_stack(vjust=0.5),
            colour = "white",
            size = 5) +
  coord_flip() +
  scale_x_discrete(limits = c("Direct", "Linker")) +
  scale_y_continuous(label = percent) +
  scale_fill_tableau(palette = "Miller Stone", 
                     direction = -1, 
                     guide = guide_legend(reverse = TRUE)) +
  theme_minimal_grid() + 
  theme(legend.position = "top") +
  labs(x = "",
       y = "Percentage of hybrids",
       fill = "")

# p <- ggplot(arms.dt[, .N, by = .(L_region, exp)], aes(x = exp, y = N, fill = L_region)) +
#   geom_col(width = 0.5, position = "fill") +
#   coord_flip() + scale_x_discrete(limits = c("Direct", "Linker")) +
#   scale_y_continuous(label = percent) +
#   scale_fill_tableau(palette = "Miller Stone", direction = -1) +
#   theme_minimal_grid() + theme(legend.position = "top") +
#   labs(x = "",
#        y = "Percentage of hybrids",
#        fill = "")

ggsave(p, filename = file.path(plot.path, "linker_nolinker_regions.pdf"), width = 7, height = 4)

p
```

```{r}
round(prop.table(table(arms.dt[exp == "Direct"]$L_region)) * 100, 1)
round(prop.table(table(arms.dt[exp == "Linker"]$L_region)) * 100, 1)
```

# Figure 2 D - Duplexes

### Linkers

```{r}
atlas.hybrids.dt <- linker.dt[, total_count := .N, by = .(L_seqnames, R_seqnames)]
atlas.hybrids.dt <- atlas.hybrids.dt[L_seqnames == R_seqnames]
atlas.hybrids.dt <- atlas.hybrids.dt[!(L_seqnames == "rRNA_45S" & R_seqnames == "rRNA_45S")]
atlas.hybrids.dt <- atlas.hybrids.dt[!(L_seqnames == "rDNA" & R_seqnames == "rDNA")]
atlas.hybrids.dt <- atlas.hybrids.dt[!(L_seqnames == "rRNA_5S" & R_seqnames == "rRNA_5S")]
atlas.hybrids.dt <- atlas.hybrids.dt[!(L_seqnames == "rRNA5S" & R_seqnames == "rRNA5S")]
atlas.hybrids.dt <- atlas.hybrids.dt[!grepl("tRNA", L_seqnames)]
atlas.hybrids.dt <- atlas.hybrids.dt[!grepl("tRNA", R_seqnames)]

linker.list <- split(atlas.hybrids.dt, by = c("L_seqnames", "R_seqnames"))
linker.list <- linker.list[elementNROWS(linker.list) > 1]
linker.clusters.list <- parallel::mclapply(linker.list, cluster_hybrids, percent_overlap = 0.5, mc.cores = 4)
tic()
linker.clusters.list <- lapply(linker.list, cluster_hybrids, percent_overlap = 0.5)
toc()
linker.clusters.dt <- rbindlist(linker.clusters.list, use.names = TRUE, fill = TRUE)
```

```{r}
dt <- data.table(exp = c("Linker", "Direct"),
                 counts = c(nrow(linker.clusters.dt[L_seqnames == R_seqnames][!L_seqnames %in% c("tRNA", "rDNA", "rRNA_5S")][grep("C", cluster), .N, by = .(cluster, L_seqnames, R_seqnames)]),
                            nrow(nolinker.dt[L_seqnames == R_seqnames][grep("C", cluster), .N, by = .(cluster, L_seqnames, R_seqnames)])))

p <- ggplot(dt, aes(x = exp, y = counts)) +
  geom_col(width = 0.5) +
  scale_fill_tableau() +
  theme_minimal_grid() + theme(legend.position = "top") +
  labs(x = "",
       y = "Number of clusters",
       fill = "")

ggsave(p, filename = "linker_nolinker_cluster_counts.pdf", width = 5, height = 7)
```

# Figure 2 E - hybridisation energy

```{r}
linker.dt <- fread("~/Dropbox (The Francis Crick)/comp_hiclip/hiclip/linker/linker.clusters.mfe.tsv.gz")

mfe.dt <- rbindlist(list(nolinker.dt[L_seqnames == R_seqnames][grep("^C", cluster)][, mean(mean_shuffled_mfe), by = .(L_seqnames, cluster)][, `:=` (exp = "Direct", mfe = "Shuffled")],
                         nolinker.dt[L_seqnames == R_seqnames][grep("^C", cluster)][, mean(mfe), by = .(L_seqnames, cluster)][, `:=` (exp = "Direct", mfe = "Duplex")],
                         linker.dt[L_seqnames == R_seqnames][grep("^C", cluster)][, mean(mean_shuffled_mfe), by = .(L_seqnames, cluster)][, `:=` (exp = "Linker", mfe = "Shuffled")],
                         linker.dt[L_seqnames == R_seqnames][grep("^C", cluster)][, mean(mfe), by = .(L_seqnames, cluster)][, `:=` (exp = "Linker", mfe = "Duplex")]))

setnames(mfe.dt, "V1", "energy")

p <- ggplot(mfe.dt, aes(x = energy, colour = exp, linetype = mfe)) +
  geom_density(size = 1) +
  scale_colour_manual(values = ht_colours) +
  theme_minimal_grid() + theme(legend.position = "top") +
  labs(x = "Hybridisation energy",
       y = "Density",
       colour = "",
       linetype = "")  

ggsave(p, filename = file.path(plot.path, "linker_nolinker_mfe.pdf"), width = 7, height = 4)

p
```

# Figure 2 D - Overlaps

```{r}
linker.clusters <- collapse_clusters(linker.dt[L_seqnames == R_seqnames][cluster_hybrid_count >= 2], mode = "wide")
nolinker.clusters <- collapse_clusters(nolinker.dt[L_seqnames == R_seqnames][cluster_hybrid_count >= 2], mode = "wide")

linker.clusters <- primavera::reorient_hybrids(linker.clusters)
nolinker.clusters <- primavera::reorient_hybrids(nolinker.clusters)

bedpe.colnames <- c("L_seqnames", "L_start", "L_end", "R_seqnames", "R_start", "R_end", "name", "count", "L_strand", "R_strand")

linker.bedpe.dt <- linker.clusters[, ..bedpe.colnames]
linker.bedpe.dt[, `:=`(
  L_start = L_start - 1,
  R_start = R_start - 1
)]
nolinker.bedpe.dt <- nolinker.clusters[, ..bedpe.colnames]
nolinker.bedpe.dt[, `:=`(
  L_start = L_start - 1,
  R_start = R_start - 1
)]

linker.bedpe <- tempfile(tmpdir = getwd(), fileext = ".bedpe")
nolinker.bedpe <- tempfile(tmpdir = getwd(), fileext = ".bedpe")
ol <- tempfile(tmpdir = getwd(), fileext = ".bedpe")

fwrite(linker.bedpe.dt, file = linker.bedpe, sep = "\t", col.names = FALSE)
fwrite(nolinker.bedpe.dt, file = nolinker.bedpe, sep = "\t", col.names = FALSE)

cmd <- paste("bedtools pairtopair -rdn -a", linker.bedpe, "-b", nolinker.bedpe, ">", ol)
system(cmd)
bedpe.dt <- fread(ol, col.names = c(paste0(bedpe.colnames, ".x"), paste0(bedpe.colnames, ".y")))

file.remove(linker.bedpe)
file.remove(nolinker.bedpe)
file.remove(ol)

upset.input <- c("Linker" = nrow(linker.clusters) - nrow(bedpe.dt),
               "Direct" = nrow(nolinker.clusters) - nrow(bedpe.dt),
               "Linker&Direct" = nrow(bedpe.dt))

pdf(file.path(plot.path, "linker_nolinker_upset.pdf"), width = 7, height = 5)
upset(fromExpression(upset.input), 
      order.by = "freq",
      mainbar.y.label = "Intersecting number of duplexes",
      sets.x.label = "Total number of duplexes")s
dev.off()
```

# # Figure 2 D - Inset - RNA abundance of overlapping genes

```{r}
ol.genes <- unique(c(bedpe.dt$L_seqnames.x, bedpe.dt$L_seqnames.y, bedpe.dt$R_seqnames.x, bedpe.dt$R_seqnames.y))
nol.genes <- unique(c(linker.bedpe.dt$L_seqnames, linker.bedpe.dt$R_seqnames, nolinker.bedpe.dt$L_seqnames, nolinker.bedpe.dt$R_seqnames))

gene.counts.dt <- fread("~/Dropbox (The Francis Crick)/comp_hiclip/4suseq/GSE84722_ST1.txt.gz")
gene.counts.dt$mean_count <- rowMeans(gene.counts.dt[, .(RZ_Mature_copies, RH_Mature_copies, PA_Mature_copies)], na.rm = TRUE)
gene.counts.dt[, ensg := tstrsplit(Gene, "\\.")[[1]]]

counts.dt <- data.table(gene = c(ol.genes, nol.genes),
                        type = c(rep("Overlap", length(ol.genes)), rep("No overlap", length(nol.genes))))
counts.dt <- counts.dt[!grepl("_", gene)]
counts.dt[, ensg := tstrsplit(tstrsplit(gene, ":")[[2]], "\\.")[[1]]]

counts.dt <- merge(counts.dt, gene.counts.dt[, .(ensg, mean_count)], by = "ensg")

p <- ggplot(counts.dt, aes(x = type, y = mean_count, fill = type)) +
  geom_boxplot(width = 0.5, outlier.shape = NA) +
  scale_y_log10() +
  annotation_logticks(sides = "l") +
  scale_fill_brewer(palette = "Accent") +
  coord_cartesian(ylim = c(1e-1, 1e3)) +
  # geom_label(data = gene.counts.dt[mean_count > 0, .N, by = .(stau)], aes(label = N, y = 0), vjust = -0.1) +
  labs(x = "",
       y = "Mean gene copy number",
       fill = "") +
  theme_minimal_grid() + theme(legend.position = "none")

ggsave(p, filename = file.path(plot.path, "overlap_rnaabundance.pdf"), width = 5, height = 4)

p
```


# Figure 2 F - Duplex span

```{r}
linker.clusters <- collapse_clusters(linker.dt[L_seqnames == R_seqnames][cluster_hybrid_count >= 2], mode = "median")
nolinker.clusters <- collapse_clusters(nolinker.dt[L_seqnames == R_seqnames][cluster_hybrid_count >= 2], mode = "median")

nolinker.clusters[, loop := R_start - L_end - 1][, exp := "Linker"]
linker.clusters[, loop := R_start - L_end - 1][, exp := "Direct"]

intragenic.clusters.dt <- rbindlist(list(nolinker.clusters, linker.clusters))[L_seqnames == R_seqnames]

p <- ggplot(intragenic.clusters.dt, aes(x = loop, colour = exp)) +
  geom_density(size = 1) +
  scale_colour_manual(values = ht_colours) +
  scale_x_log10(label = comma) + annotation_logticks() +
  theme_minimal_grid() + theme(legend.position = "top") +
  labs(x = "Duplex span",
       y = "Density",
       colour = "",
       linetype = "")  

ggsave(p, filename = file.path(plot.path, "intragenic_linker_nolinker_duplex_span.pdf"), width = 5, height = 4)

p
```

